File: all/2785_sort_vowels_in_a_string/sol_counting_sort.py
NON_DECREASING_VOWEL_CHARACTERS_STRING = 'AEIOUaeiou'
VOWEL_CHARACTERS_SET = set(NON_DECREASING_VOWEL_CHARACTERS_STRING)


class Solution:
    def sortVowels(self, s: str) -> str:
        result_characters_list = list(s)
        vowel_character_count_dict = {vowel: 0 for vowel in 'AEIOUaeiou'}

        for c in s:
            if c not in VOWEL_CHARACTERS_SET:
                continue

            vowel_character_count_dict[c] += 1

        i = 0

        current_vowel_character_index = 0

        for i in range(len(result_characters_list)):
            if (result_characters_list[i]) not in VOWEL_CHARACTERS_SET:
                continue

            while (
                0
                ==
                (
                    vowel_character_count_dict[
                        NON_DECREASING_VOWEL_CHARACTERS_STRING[
                            current_vowel_character_index
                        ]
                    ]
                )
            ):
                current_vowel_character_index += 1

            current_vowel_character = (
                NON_DECREASING_VOWEL_CHARACTERS_STRING[
                    current_vowel_character_index
                ]
            )

            result_characters_list[i] = current_vowel_character
            (vowel_character_count_dict[current_vowel_character]) -= 1

        return ''.join(result_characters_list)


if '__main__' == __name__:
    sol = Solution()
    testcases_list = [
        ['lEetcOde'],
        ['lYmpH'],
    ]

    for testcase in testcases_list:
        print(sol.sortVowels(*testcase))

File: all/2785_sort_vowels_in_a_string/sol_counting_sort.c
#include <stdbool.h>

#define VOWEL_CHARACTERS_COUNT (10)

bool isVowel(const char c) {
    switch (c) {
        case 'a':
        case 'e':
        case 'i':
        case 'o':
        case 'u':
        case 'A':
        case 'E':
        case 'I':
        case 'O':
        case 'U':
            return true;

            break;
        default:
            break;
    }

    return false;
}

int vowelsToIndexMapping(const char vowel) {
    switch (vowel) {
        case 'A':
            return 0;
        case 'E':
            return 1;
        case 'I':
            return 2;
        case 'O':
            return 3;
        case 'U':
            return 4;
        case 'a':
            return 5;
        case 'e':
            return 6;
        case 'i':
            return 7;
        case 'o':
            return 8;
        case 'u':
            return 9;
        default:
            break;
    }

    return -1;
}

char* sortVowels(char* s) {
    const char* NON_DECREASING_VOWEL_CHARACTERS_STRING = "AEIOUaeiou";
    const char vowels_order_array[] = {
        'A', 'E', 'I', 'O', 'U', 'a', 'e', 'i', 'o', 'u',
    };
    int vowels_counter[VOWEL_CHARACTERS_COUNT] = {0};
    int next_character_index = 0;

    for (int i = 0; '\0' != (s[i]); ++i) {
        if (!isVowel(s[i])) {
            continue;
        }

        ++vowels_counter[vowelsToIndexMapping(s[i])];
    }

    for (int i = 0, current_vowel_character_index = 0; '\0' != (s[i]); ++i) {
        if (!isVowel(s[i])) {
            continue;
        }

        while (0 == vowels_counter[current_vowel_character_index]) {
            ++current_vowel_character_index;
        }

        s[i] = (
            NON_DECREASING_VOWEL_CHARACTERS_STRING[
                current_vowel_character_index
            ]
        );
        --(vowels_counter[current_vowel_character_index]);
    }

    return s;
}
